src/librustc/hir/check_attr.rs

```diff
@@ -154,7 +154,59 @@ impl<'a, 'tcx> CheckAttrVisitor<'a, 'tcx> {
         let hints: Vec<_> = item.attrs
             .iter()
             .filter(|attr| attr.name() == "repr")
-            .filter_map(|attr| attr.meta_item_list())
+            .filter_map(|attr| {
+                let list = attr.meta_item_list();
+
+                // Emit warnings with `repr` either has a literal assignment (`#[repr = "C"]`) or
+                // no hints (``#[repr]`)
+                let has_hints = list.as_ref().map(|ref list| !list.is_empty()).unwrap_or(false);
+                if !has_hints {
+                    let mut suggested = false;
+                    let mut warn = if let Some(ref lit) = attr.value_str() {
+                        // avoid warning about empty `repr` on `#[repr = "foo"]`
+                        let sp = match format!("{}", lit).as_ref() {
+                            "C" | "packed" | "rust" | "u*" | "i*" => {
+                                let lo = attr.span.lo() + BytePos(2);
+                                let hi = attr.span.hi() - BytePos(1);
```

Comment:
So, this skips the leading `#[` and trailing `]`, but only for some cases (e.g., `repr = "C"` but not `repr = "B"`). What's the rationale for that extra work and inconsistency? Why not always use `attr.span`?